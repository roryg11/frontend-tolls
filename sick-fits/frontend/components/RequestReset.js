import React, { Component } from 'react';
import { Mutation } from "react-apollo";
import gql from "graphql-tag";
import Form from "./styles/Form"; 
import Error from "./ErrorMessage";

const REQUEST_RESET_MUTATION = gql`
    mutation REQUEST_RESET_MUTATION(
        $email: String!
    ){
        requestReset(email: $email){
            message
        }
    }
`

class RequestReset extends Component {
    state = {
        email: "",
    }

    stateChange = (e) => {
        e.preventDefault();
        const prop = e.target.name; 
        this.setState({
            [prop]: e.target.value
        });
    }

    render() {
        return (
            <Mutation mutation={REQUEST_RESET_MUTATION} 
                      variables={this.state}>
                { (requestReset, {error, loading, called})=> {
                        if(error) return <Error error={error}/>
                        if(called) return <p>Check your email for a reset password link!</p>
                        return (<Form method="post" 
                        data-test="form"
                        onSubmit={ async e =>{
                            e.preventDefault();
                            await requestReset();
                            this.setState({email: ""});
                        }}>
                            <fieldset>
                                <label>
                                    Email
                                    <input 
                                    type="text" 
                                    name="email" 
                                    placeholder="Email"
                                    value={this.state.email}
                                    onChange={this.stateChange}/>
                                </label>
                            </fieldset>
                            <button>Request Password Reset</button>
                        </Form>)
                    }
                }
            </Mutation>
            
        );
    }
}

export default RequestReset;
export { REQUEST_RESET_MUTATION };