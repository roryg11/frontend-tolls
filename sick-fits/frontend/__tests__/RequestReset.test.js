import {shallow, mount} from "enzyme";
import toJSON from "enzyme-to-json";
import { MockedProvider } from "react-apollo/test-utils";
import wait from "waait";
import RequestReset, { REQUEST_RESET_MUTATION }  from "../components/RequestReset";
import { fakeUser } from "../lib/testUtils";

const mocks = [
    {
        request: {
            query: REQUEST_RESET_MUTATION,
            variables: {
                email: "rory@gmail.com"
            }
        },
        result: { 
            data: {
                requestReset: { message: "success", __typename: 'Message'}
            }
        }
        
    }
];

describe('<RequestReset/>', ()=>{
    it('renders and matches snapshot', async ()=>{
        const wrapper = mount(
            <MockedProvider>
                <RequestReset/>
            </MockedProvider>
        );

        await wait();
        wrapper.update();

        const form = wrapper.find('form');
        expect(toJSON(form)).toMatchSnapshot();
    });

    it('calls the reset mutation', async ()=>{
        const wrapper = mount(
            <MockedProvider mocks={mocks}>
                <RequestReset/>
            </MockedProvider>
        );

        await wait();
        wrapper.update();

        wrapper.find('input').simulate('change', { target: { name: "email", value: "rory@gmail.com"}});
        wrapper.find('form').simulate('submit');
        await wait();
        wrapper.update();

        const message = wrapper.find('p');
        expect(message.text()).toContain("Check your email for a reset password link!");
    });
});