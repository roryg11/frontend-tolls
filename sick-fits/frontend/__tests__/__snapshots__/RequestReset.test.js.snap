// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestReset/> renders and matches snapshot 1`] = `
<form
  className="Form-sc-1xszr8q-0 krRxky"
  data-test="form"
  method="post"
  onSubmit={[Function]}
>
  <fieldset>
    <label>
      Email
      <input
        name="email"
        onChange={[Function]}
        placeholder="Email"
        type="text"
        value=""
      />
    </label>
  </fieldset>
  <button>
    Request Password Reset
  </button>
</form>
`;
